File: java-app/src/main/java/com/showcase/java/domain/projection/OverdueTasksProjection.java
package com.showcase.java.domain.projection;

import com.showcase.java.domain.event.DomainEvent;
import com.showcase.java.domain.model.TaskStatus;
import com.showcase.java.domain.model.TaskSummary;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class OverdueTasksProjection implements Projection {

    private final ConcurrentHashMap<String, TaskSummary> tasks = new ConcurrentHashMap<>();

    @Override
    public void handle(DomainEvent event) {
        switch (event) {
            case DomainEvent.TaskCreated e -> tasks.put(e.aggregateId(), new TaskSummary(
                    e.aggregateId(),
                    e.title(),
                    new TaskStatus.Open(),
                    e.priority(),
                    null,
                    e.dueDate(),
                    e.tags()
            ));
            case DomainEvent.TaskAssigned e -> tasks.computeIfPresent(e.aggregateId(),
                    (_, s) -> s.withAssignee(e.assigneeId())
                               .withStatus(new TaskStatus.InProgress(e.assigneeId())));
            case DomainEvent.TaskPriorityChanged e -> tasks.computeIfPresent(e.aggregateId(),
                    (_, s) -> s.withPriority(e.newPriority()));
            case DomainEvent.TaskCompleted e -> tasks.computeIfPresent(e.aggregateId(),
                    (_, s) -> s.withStatus(new TaskStatus.Completed(e.completedAt())));
            case DomainEvent.TagAddedToTask e -> tasks.computeIfPresent(e.aggregateId(),
                    (_, s) -> {
                        var newTags = new HashSet<>(s.tags());
                        newTags.add(e.tag());
                        return s.withTags(newTags);
                    });
            case DomainEvent.TaskMovedToProject _ -> {
                // Project move doesn't affect overdue tracking
            }
        }
    }

    public List<TaskSummary> getOverdueTasks(LocalDate asOf) {
        return tasks.values().stream()
                .filter(s -> s.dueDate() != null
                        && s.dueDate().isBefore(asOf)
                        && !(s.status() instanceof TaskStatus.Completed))
                .toList();
    }

    public List<TaskSummary> getOverdueTasks() {
        return getOverdueTasks(LocalDate.now());
    }
}

File: kotlin-app/src/main/kotlin/com/showcase/kotlin/domain/projection/OverdueTasksProjection.kt
package com.showcase.kotlin.domain.projection

import com.showcase.kotlin.domain.event.DomainEvent
import com.showcase.kotlin.domain.model.TaskStatus
import com.showcase.kotlin.domain.model.TaskSummary
import java.time.LocalDate
import java.util.concurrent.ConcurrentHashMap

class OverdueTasksProjection : Projection {
    private val tasks = ConcurrentHashMap<String, TaskSummary>()

    override suspend fun handle(event: DomainEvent) {
        when (event) {
            is DomainEvent.TaskCreated -> tasks[event.aggregateId] = TaskSummary(
                id = event.aggregateId,
                title = event.title,
                status = TaskStatus.Open,
                priority = event.priority,
                assignee = null,
                dueDate = event.dueDate,
                tags = event.tags
            )
            is DomainEvent.TaskAssigned -> tasks.computeIfPresent(event.aggregateId) { _, summary ->
                summary.copy(
                    assignee = event.assigneeId,
                    status = TaskStatus.InProgress(event.assigneeId)
                )
            }
            is DomainEvent.TaskPriorityChanged -> tasks.computeIfPresent(event.aggregateId) { _, summary ->
                summary.copy(priority = event.newPriority)
            }
            is DomainEvent.TaskCompleted -> tasks.computeIfPresent(event.aggregateId) { _, summary ->
                summary.copy(status = TaskStatus.Completed(event.completedAt))
            }
            is DomainEvent.TagAddedToTask -> tasks.computeIfPresent(event.aggregateId) { _, summary ->
                summary.copy(tags = summary.tags + event.tag)
            }
            is DomainEvent.TaskMovedToProject -> {
                // Project move doesn't affect overdue tracking
            }
        }
    }

    fun getOverdueTasks(asOf: LocalDate = LocalDate.now()): List<TaskSummary> =
        tasks.values.filter { summary ->
            summary.dueDate != null &&
                summary.dueDate.isBefore(asOf) &&
                summary.status !is TaskStatus.Completed
        }
}
